feat(pipeline)Data Model for Visualization and utility function

abhinandan13jan opened the pull request:
@karthikjeeyar Can you pull and see if this will work for your implementation
Run the install_mock_pipelines.sh script from mock area to add the new tasks.

Please overlook the implementation of PipelineVisualization.jsx which will be reimplemented by Karthik
I get a look like this.

![Screenshot from 2019-05-10 21-13-21](https://user-images.githubusercontent.com/24852534/57539891-1b66e800-7369-11e9-874d-34b59049993f.png)

![Screenshot from 2019-05-10 21-13-30](https://user-images.githubusercontent.com/24852534/57539900-1f930580-7369-11e9-8c18-e8e51815e63f.png)

@christian Many of the changes suggested have been covered in the latest commit, maybe Karthik can make the remaining changes in his part of implementation or I could see into it next week :)

Commit:
feat(pipeline)Data Model and utility function

christianvogt commented on frontend/public/extend/devconsole/components/pipelines/PipelineDetails.tsx:
```diff
@@ -9,17 +11,7 @@ const PipelineDetails = ({ obj: pipeline }) => (
-            );
-          })}
-        </dl>
+        <PipelineVisualization pipeline={pipeline} />
```
Keep the list of tasks for now.
Move this to a row above the title "Pipeline Overview".

abhinandan13jan replied:
done

christianvogt commented on frontend/public/extend/devconsole/shared/components/pipelineVisualization.tsx:
```diff
@@ -0,0 +1,68 @@
+    this.state = {
+      tasks:[]
+   }
+    this.setState({ tasks: getOutput(this.props.pipeline.spec.tasks) });
```
Do not set this as state here as the state is never used.

abhinandan13jan replied:
removed

christianvogt commented on frontend/public/extend/devconsole/shared/components/pipelineVisualization.tsx:
```diff
@@ -0,0 +1,68 @@
+  }
+}
+
+export default PipelineVisualization;
```
I get that this is just a temp solution but it should be formatted and follow good conventions.
Pascal case the file name and move it to our components folder.

abhinandan13jan replied:
Pascal cased file and is in devconsole/shared/component, will be placed after @karthikjeeyar's implementation

christianvogt commented on frontend/public/extend/devconsole/shared/components/pipelineVisualization.tsx:
```diff
@@ -0,0 +1,68 @@
+}
+interface PipelineVisualizationStates {
+  tasks: any;
+}
```
Remove the state and add a proper type to the props.

abhinandan13jan replied:
this will be done in @karthikjeeyar's implementation

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,298 @@
+  return output;
+};
+
+export const getOutput = (actualTasks):Array<Array<Object>> => {
```
Give a more descriptive name than `getOutput`.
Use the convention `string[]` over `Array<string>`. Especially when declaring two dimensional arrays `string[][]`.
That being said this needs a proper type. The entire data structure should be typed.

abhinandan13jan replied:
addressed

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,298 @@
+  tasks.map((task) => {
+    if (task.runAfter && task.runAfter.length > 0) {
+      //do nothing
+    } else {
```
Make sure to address this in the final implementation.
Invert the condition and use a single else.
Do not leave empty blocks.

abhinandan13jan replied:
addressed

christianvogt reviewed (commented):
Just a few comments as I know this is still a `WIP`.
Don't forget to format with prettier.

Commit:
feat(visualization): pipeline visual data model and test

Commit:
feat(visualization): pipeline visual data model and test

Commit:
feat(visualization): pipeline visual data model and test

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+  return output;
+};
+
+export const getPipelineTasks = (actualTasks: Task[]): Task[][] => {
```
Not sure what `actualTasks` means to a user of the api.
I suggest changing it to simply `tasks`.

abhinandan13jan replied:
changed to taskList

christianvogt commented on frontend/public/extend/devconsole/shared/components/PipelineVisualization.tsx:
```diff
@@ -0,0 +1,115 @@
+  }
+}
+
+export default PipelineVisualization;
```
Move file to `devconsole/components/pipelines`

abhinandan13jan replied:
this component will be added in @karthikjeeyar 's PR

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+    return;
+  });
+  return out;
+};
```
Move file to `devconsole/utils/pipeline-utils.ts`

abhinandan13jan replied:
moved

christianvogt commented on frontend/public/models/index.ts:
```diff
@@ -211,6 +211,32 @@ export const PipelinerunModel: K8sKind = {
+  id: 'taskRun',
+  labelPlural: 'TaskRuns',
+};
+
```
We should keep our models in our own repo / folder.

abhinandan13jan replied:
will be taken care by Karthik's PR. However k8s actions don't seem to work with models in our repo.

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+};
+
+export const getPipelineTasks = (actualTasks: Task[]): Task[][] => {
+  //P.S: Each element in out array is referred to as stage
```
Fix sentence structure.

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+  for (let i = 0; i < output.length; i++) {
+    if (conditions.hasRunAfterDependency(output[i])) {
+      let flag = -1;
+      for (let j = 0; j < output.length; j++) {
```
`j` should start as `j = i + 1` in all cases.
Then you can remove your check for `i < j`

abhinandan13jan replied:
done

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+        const temp = output[flag];
+        output[flag] = output[i];
+        output[i] = temp;
+      }
```
Move this if condition to the end of the first `for` loop as it's duplicate logic seen below.

abhinandan13jan replied:
moved

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+  const tasks = sortTasksByRunAfterAndFrom(actualTasks);
+
+  //Step 2: Push all nodes without any dependencies in different stages
+  tasks.map((task) => {
```
```suggestion
  tasks.forEach((task) => {
```
Then you can remove your empty `return` statement.

abhinandan13jan replied:
done

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+  });
+
+  //Step 3: Push nodes with 'from' dependency and stack similar tasks in a stage
+  tasks.map((task) => {
```
```suggestion
  tasks.forEach((task) => {
```

abhinandan13jan replied:
done

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+    if (!conditions.hasRunAfterDependency(task) && conditions.hasFromDependency(task)) {
+      let flag = out.length - 1;
+      for (let i = 0; i < out.length; i++) {
+        out[i].map((t) => {
```
```suggestion
        out[i].forEach((t) => {
```

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+  });
+
+  //Step 4: Push nodes with 'runAfter' dependencies and stack similar tasks in a stage
+  tasks.map((task) => {
```
```suggestion
  tasks.forEach((task) => {
```

abhinandan13jan replied:
done

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+    if (conditions.hasRunAfterDependency(task)) {
+      let flag = out.length - 1;
+      for (let i = 0; i < out.length; i++) {
+        out[i].map((t) => {
```
```suggestion
        out[i].forEach((t) => {
```

abhinandan13jan replied:
done

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+
+  //Step 3: Push nodes with 'from' dependency and stack similar tasks in a stage
+  tasks.map((task) => {
+    if (!conditions.hasRunAfterDependency(task) && conditions.hasFromDependency(task)) {
```
Is it necessary to check both of these conditions? Why not just `hasFromDependency` ?

abhinandan13jan replied:
yes, we need to keep the condition 'hasRunAfterDependency' as a task may have both 'runAfter' and 'from' dependency and in that case it must be checked in next 'loop' after all the from dependencies are in place.

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+      const nextToFlag = out[flag + 1] ? out[flag + 1] : null;
+      if (
+        nextToFlag &&
+        nextToFlag[0].resources.inputs[0].from[0] === task.resources.inputs[0].from[0]
```
Is it safe to do `nextToFlag[0].resources.inputs[0].from[0]` without checking if the `nextToFlag` task has the same structure?

abhinandan13jan replied:
added safety checks

christianvogt commented on frontend/public/extend/devconsole/shared/utils/visualization.ts:
```diff
@@ -0,0 +1,125 @@
+      }
+      const nextToFlag = out[flag + 1] ? out[flag + 1] : null;
+
+      if (nextToFlag && nextToFlag[0].runAfter[0] === task.runAfter[0]) {
```
Is it safe to do `nextToFlag[0].runAfter[0]` without checking if the `nextToFlag` task has the same structure?

abhinandan13jan replied:
added safety checks

christianvogt reviewed (commented):
missing unit tests.

Commit:
feat(pipeline): data model for pipeline visualization

abhinandan13jan commented:
@christianvogt Made changes, added safety checks, checking from i+1, added unit tests, we need to keep the condition 'hasRunAfterDependency' as a task may have both 'runAfter' and 'from' dependency and in that case it must be checked in next 'loop' after all the from dependencies are in place.

abhinandan13jan commented:
@christianvogt @karthikjeeyar Closing this PR as all changes pertaining to visualization will come in @karthikjeeyar 's PR
